### ISSUE TITLE
Securing lambda secrets with KMS

### ISSUE THREAD

**Issue by rarkins** [serverless/serverless#472]
I would like to discuss how to leverage KMS capabilities to secure secrets needed by serverless modules.

An example requirement is securing a token/password for an outbound mail server.

A more complex requirement is securing an encryption key that's used for symmetric encryption and decryption of per-user credentials in a database (e.g. OAuth tokens from a third party service are encrypted before storage into DynamoDB).

The current approach in serverless (I believe) is that these secrets are supplied to lambda via `.env` files. i.e. it requires the person deploying them or some form of automated deployment system to keep these secrets locally, so they can be included at deploy time. Securing them on the local or deployment systems is then a separate challenge.

**Comment by rarkins** [comment]
[AWS's Key Management Service](https://aws.amazon.com/kms/) provides some great capabilities that could be leveraged by the serverless project to avoid the [turtles all the way down](http://technologyadvice.github.io/lock-up-your-customer-accounts-give-away-the-key/) security scenario.

To summarise, there are three approaches that could be taken for securing secrets:

- For secrets less than 4KB in size, KMS can use symmetric encryption to encrypt/decrypt
- KMS can be used to generate or secure a project key that is used to decrypt all secrets (mail credentials, database passwords, etc)
- KMS can even use a unique key to encrypt every secret, and use envelope encryption in order to decrypt

Like you'd expect, usage of KMS is policed using IAM roles and user policies. i.e. access to a 'master key' for production could be limited to only the production lambda functions plus the administrator account(s).

**Comment by rarkins** [comment]
Here's the simplest way that KMS could be used to better secure secrets in serverless projects:

- A customer master key (CMK) is generated in KMS, with access only to production lambda functions (an IAM role) and the administrator account
- Any static secrets (e.g. DB password, mail credentials) are manually encrypted by the project administrator using KMS and the encrypted values are stored as values in `.env` files
- Such `.env` files containing encrypted secrets as as well as any unencrypted non-secrets (e.g. DB user, DB host, etc) can then even be committed to the repo, if desired
- lambda functions are modified so that they use the KMS decrypt function (part of AWS SDK) on secrets prior to use (e.g. prior to generating the DB connection string), ensuring unencrypted secrets are stored only in memory
- If lambda functions dynamically write and read secrets to databases (e.g. storing OAuth tokens or SSH keys in DynamoDB) then they could be modified to wrap such calls in encrypt/decrypt calls to KMS

In the above scenario, it removes the need for administrators or automated deploy systems to keep unencrypted secrets/credentials stored on disk in their local serverless project. In fact, administrators could/should even remove their KMS access rights except when encrypting static keys, so that an attacker compromising their IAM credentials can't simply call the AWS SDK to decrypt all `.env` files too. 

This scenario perhaps doesn't even need any serverless support that I can think of, i.e. developers could do all the above today without modifications to source code, however they'd need to set up KMS manually. Unfortunately though, it's perhaps not ideally suited to the stateless lambda approach. Our testing has shown that the latency to decrypt a secret using a KMS CMK is often around 2 seconds, which is too high for most lambda functions, especially if it needs to be invoked regularly. So although this approach would work quite well on a server-based node.js project where the secrets can be decrypted at app init time and reused, it might be a latency and/or cost burden on serverless projects.

**Comment by rarkins** [comment]
A second way KMS could be used would be to get more involved in the serverless deployment workflow, and decrypt secrets at the time of deployment, through the use of envelope encryption.

This approach is quite similar to that described for [Cryptex](https://github.com/TechnologyAdvice/Cryptex) as described in [Lock Up Your Customer Accounts, Give Away The Key](http://technologyadvice.github.io/lock-up-your-customer-accounts-give-away-the-key/), using the 'IAM Roles' approach described there.

As described above, the achilles heel for using KMS directly in serverless functions is the KMS latency. So instead here I will describe a way that serverless functions can have direct access to unencrypted secrets/keys while still avoiding the turtles problem of where to store the 'master' key.

Instead of using the KMS CMK to directly encrypt/decrypt, we use it to store our own encryption key that is used to decrypt env files. In theory though you could use one key per env file, or even one key per encrypted field, if there were advantages to that. For now I'll assume the simplest option of one encryption key per stage per project, protected/stored by KMS.

At a high level:

- A data key per stage is created and envelope encrypted using the KMS CMK for that stage
- A special KMS lambda function (provided as part of serverless) is provisioned, and granted access to the relevant CMK for that stage
- The envelope key is stored in the project somehow
- Secrets are encrypted using the raw data key and their values stored in env files
- Lambda functions are pushed to S3 like normal, including env files with optionally encrypted values
- After pushing to S3, but before deploying to lambda, the "special KMS lambda function" is invoked and it reads the new zip from S3, decrypts secrets in env files using KMS, and then overwrites the existing env files with replacements containing decrypted values
- The updated zip files are deployed to lambda

i.e. this achieves the following outcomes:

- Secrets in env files can be encrypted in such a way that there is no direct threat from them being leaked
- Admins or deploy systems no longer need to keep unencrypted secrets on hand
- Lambda functions get direct access to raw keys/secrets and don't have any latency from decrypting them every invocation

The impact to the serverless lambda deploy process is fairly large though. I can think of two possible ways for the 'special' lambda function to be invoked:

1. It's invoked automatically based on the S3 upload event, *and* it performs an automatic deploy after decrypting
2. It's invoked via API or SDK and can be inserted in the "chain" of deploy events after upload and before actual deploy

**Comment by rarkins** [comment]
I will leave this hear for now, awaiting feedback and ideas if this is worthwhile to pursue as part of the serverless project.

**Comment by ryansb** [comment]
@rarkins Yes, KMS is a solid choice for secret (or even semi-secret) environment management. I don't have time to respond to all this right now, but it's something I've been thinking about too. For now, I'm working on more language support.

**Comment by rarkins** [comment]
@eahefnawy why is this closed without explanation? The Meta Sync approach is a rather insecure hack and surely can't be "the final solution" when it comes to serverless secret management.

**Comment by eahefnawy** [comment]
oh really sorry @rarkins ... I just close outdated/inactive issues to cleanup our backlog :) ... Closing the issue only indicates that we're not actively working on this but anyone is free to pitch in with more discussion. But as always, thanks a lot for your contributions :blush:

**Comment by rarkins** [comment]
Hi @eahefnawy, thanks for clarifying. I'd like to discuss this further from a meta level.

GitHub Issues is not the world's best project management tool, but closing issues just for being inactive doesn't seem long-term beneficial to the project. Have you considered a Kanban-like approach such as @waffleio where the default state for an issue is "Backlog", triaged issues are "Ready", etc. If the challenge you're trying to address is "clutter" and distinguishing new/incoming issues from inactive ones like this, then you could also add a label to indicate "not any time soon". I use a label/column "Deferred" in my own projects to mean that - it exists between Backlog and Ready. At the very, very lease - could you use a label "wontdo" or similar so it's clear which issues are completed and which simply got bumped off?

Closing issues in the manner you're doing will build a negative cycle in the "community" (assuming the goal is to build an active one..), e.g.:

 - Someone new to serverless in the future who thinks "wow, there's a real hole in secrets management" is less likely to comment on this relevant issue if they see it's closed
 - For someone like me, I'm much less likely to raise issues in future if it's the project policy to close them off

**Comment by eahefnawy** [comment]
@rarkins you're absolutely right, we're really sorry :disappointed: . We've been trying to keep our focus on the road map we had in Trello, but then we decided to manage our backlog completely on Github cause this is where our users are (we noticed that we don't pay attention to Github Issues when we used Trello). But you're right! Instead of closing we should definitely organize our labels better to reflect the **status** not the **category** of an issue. I also didn't keep in mind the fact that closing a "discussion" issue makes it less accessible for future users, since it's unlikely that someone will click on the "Closed" tap.

Thanks a lot for your constructive feedback. Please keep it coming :blush:

**Comment by ErikEvenson** [comment]
So from this discussion, I take it that the current best practice for providing, say, database credentials to a serverless function is to:

- save the credentials in `/_meta/variables/s-variables-common.json` (or something more specific)
- add an environment variable in say `s-function.json`
- get that environment variable with `process.env[key]` in the function

**Comment by eahefnawy** [comment]
@ErikEvenson Yep that should work :)

**Comment by flomotlik** [comment]
We're currently looking into secret management in the next version of serverless, but currently don't plan to integrate with KMS. Mainly this is due to KMS not being that widely used amongst our users and the amount of work we would have to put into making KMS work.

I do want to come back to this at some point in the future once we're ready to build a larger integration of private key handling. I will close the issue for now until we're returning back to KMS.

**Comment by flomotlik** [comment]
Had a conversation with @shortjared about his project https://github.com/trek10inc/serverless-secrets and the KMS integration for Serverless and we want to explore this more now

**Comment by mfulleratlassian** [comment]
Just wanted to add that I have released a plugin I am using to use KMS for variables: https://www.npmjs.com/package/serverless-plugin-kmsvariables

**Comment by shortjared** [comment]
@mfulleratlassian I love the clean management CLI of your plugin!

One of the biggest push backs we are getting from clients on secrets management, is that they don't ever want unencrypted strings in the build artifacts. If you decrypt at runtime, even someone with rights to download the code off lambda still can't get at the actual secrets. All bets are off if someone can directly push code to lambda functions with KMS decryption roles though. (But it's a manageable problem at that point.)

Just goes to show this is a good discussion to really have, and there is obviously a demand if different parties are trying to solve it.

**Comment by mfulleratlassian** [comment]
Yeah, my immediate focus was getting the variable JSON files back into our git repo. The meta sync plugin just doesn't work well for my team.

I also would love to leave the variables encrypted in the build but that would require updating the core serverless handler that loads the environment variables at lambda deploy time to use KMS directly and remove the decryption hook on functionDeploy in my plugin.

**Comment by shortjared** [comment]
Yep, `_meta` back in the repo was exactly my focus as well!

It certainly does require the overwrite of the handler, which is exactly what I did as a post lambda package hook. Just overwrite the default handler with a handler that does decryption once on cold start, and it caches for the rest of the uses.

@mfulleratlassian We were discussing perhaps a secrets plugin that handles multiple "providers". Credstash compatible, directly through kms, etc. One question I have for you, if running decryption at run time was an option, is it worth the cold start performance penalty to you (~1 second) or would you still prefer decryption right before deploy?

**Comment by flomotlik** [comment]
Now that we've talked more about secrets management I really think we should have secrets management integrated into the framework. We might ship it as a plugin in V1 first (though I'd love to have it backed in from the beginning), but its definitely very important to do.

**Comment by mfulleratlassian** [comment]
Decryption at run time is great. The 1 second hit on cold start would be worth the impact in my opinion.

**Comment by hassankhan** [comment]
We've been using [Serverless Secrets](https://github.com/trek10inc/serverless-secrets) with KMS and committing our `_meta` folder. Nothing else to add other than it would be nice to see this addressed 😄

**Comment by flomotlik** [comment]
@hassankhan yup Serverless Secrets is awesome and one of the ways we want to support encryption (or getting variables into production in general).

**Comment by satyrius** [comment]
@flomotlik is there any progress with secrets management implementation? How could we help you guys to push this to closest release?

**Comment by flomotlik** [comment]
@satyrius we do have plans how this will be implemented, but we're waiting on AWS with this specific issue. Its a super important feature together with Environment Variables.

**Comment by satyrius** [comment]
@flomotlik whats wrong on AWS? I didn's found this in the topic.

**Comment by aliatsis** [comment]
@flomotlik is there any timeline for this project? This is somewhat blocking me from updating to 1.0 from 0.5.x where I've built my own custom solution. Any update would be greatly appreciated!

**Comment by flomotlik** [comment]
I'm going to close this issue here as we're working with AWS to get a solution in place and will open up another issue as soon as we start the work on this. This is a super high priority feature for us and we'll support it as soon as AWS moves into Environment variables for lambdas.

**Comment by SydOps** [comment]
@hassankhan 

I don't really get how `serverless-secrets` works. 

I currently look for a solution to hide the secret, Will `serverless-secrets` have the lookup function that I can refer in serverless.yml?

```
provider:
  name: aws
  runtime: nodejs4.3
  stage: dev
  region: us-east-1
  environment:
    CLIENT_ID: "{{ loookup 'client_id' by kms_key }}"
    CLIENT_SECRET:  "{{ loookup 'client_secret' by kms_key }}"
```

**Comment by fernando-mc** [comment]
I've read through the entire chain here and I'm still not sure if there is a supported solution to resolve the issues laid out here. I've also opened a question on the forums here - http://forum.serverless.com/t/how-to-deal-with-environment-variables-source-control-serverless-yml/2248

I did see that the documentation has some mention of KMS but seems like the current solution is to use community plugins for something that seems like a blocker to put serverless.yml in version control with environment variables. I can understand the delay as this could possibly be a decent-sized project.

One possible architecture here is to have a secrets command built in that allows you to encrypt with KMS. The functionality would need to:

1. Create and use a specific KMS key probably defined by name in serverless.yml
2. Enable some cli functionality to allow for a workflow where you could use `serverless encrypt SOME_SECRET` Then output `ENCRYPTED_CIPHERTEXT` and add the ciphertext back into serverless.yml as an environment variable.
3. Finally, to make these values usable you'd either need to
   a) Decrypt them at deployment and build the environment variables in with the build artifacts (less secure and probably not ideal, but it does save development time) OR
   b) Give your functions KMS IAM permissions to the key and manually make the calls to KMS in your functions to decrypt the environment variables. (More overhead time in the actual function calls, but probably more secure)

I'm hoping that all this was already thought through by someone else far smarter who has taken the time to develop it. 😁 Mainly I'm wondering what the current state of this is and working from there. I realize the framework isn't all-in with AWS so this complicates things a bit as the solution I provided only deals with that.

Maybe the eventual solution is just to have a few supported plugins for this purpose for each IaaS provider?

**Comment by rarkins** [comment]
I'm not using serverless currently but if I were, I'd hope to find support for http://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html somehow instead of direct KMS access

**Comment by fernando-mc** [comment]
@rarkins Seems like the exact thing we want, but unclear if it integrates with Lambda?

I think AWS has built-in support for KMS encryption of environment variables in Lambda that decrypt at runtime so maybe the better option is to have a way to store those values and still have it decrypt at runtime somehow?

**Comment by candlerb** [comment]
Just a couple of notes.

Inside the AWS Lambda console, you have the ability to encrypt individual environment variables, using a KMS key of your choice.  They then give you some code you can paste in to your application to do the decryption.  For example, if it's python, they give you:

~~~
import boto3
import os

from base64 import b64decode

ENCRYPTED = os.environ['DATABASE_URL']
# Decrypt code should run once and variables stored outside of the function
# handler so that these are decrypted once per container
DECRYPTED = boto3.client('kms').decrypt(CiphertextBlob=b64decode(ENCRYPTED))['Plaintext']

def lambda_handler(event, context):
  # TODO handle the event here
~~~

Now, CloudFormation itself doesn't support encryption of individual keys - https://github.com/awslabs/serverless-application-model/issues/48 . Some people have done it using a lambda with NoEcho parameters.

But it seems to me that if you're storing the plaintext keys on your workstation, that's the weakest point anyway.  Many times people have accidentally committed keys into repos.

So an approach you could take instead is:

* Encrypt the key locally, using `aws kms encrypt --key-id arn:aws:kms:... --plaintext "..."`
* Paste the base64 ciphertext into your environment settings
* Use Amazon's decryption code in your lambda function (doing the decryption outside of the handler, so it's only done once at startup time)

Then the plaintext is never stored on your workstation, nor in any build artefacts.

The only problem is that accessing KMS has to be done over the Internet.  If your lambda is running in a VPC, and your VPC doesn't have a NAT gateway, then it just [times out](https://stackoverflow.com/questions/42114619/how-to-use-encrypted-environment-variables-in-aws-lambda).

**Comment by lorengordon** [comment]
What I've taken to doing is using the SSM Parameter Store to hold encrypted values as a SecureString, passing the path to the parameter to the lambda function via an ENV, giving the lambda function an IAM role with privs to the SSM parameter path and the KMS key, and retrieving/decrypting the value inside the function.

Works pretty well. The user env does not need to contain the secrets, nor are the secrets exposed to lambda or cloudformation (outside the in-memory context of an executing function).

Save the parameter:

```
aws ssm put-parameter --name /foo/bar --value bar --type SecureString
```

serverless.yml would look something like this:

```
plugins:
  - serverless-cf-vars  # mostly to make it easier to reference pseudo parameters in iam roles

custom:
  ssmFoo: /foo/bar
  ssmKms: kms-key-id

provider:
  name: aws
  runtime: nodjs6.10

  iamRoleStatements:
    - Effect: "Allow"
      Action:
        - ssm:GetParameter
        - ssm:GetParameters
      Resource:
        # Account for both regular and path-based SSM Parameters
        - "arn:#{AWS::Partition}:ssm:#{AWS::Region}:#{AWS::AccountId}:parameter/${self:custom.ssmFoo}"
        - "arn:#{AWS::Partition}:ssm:#{AWS::Region}:#{AWS::AccountId}:parameter${self:custom.ssmFoo}"
    - Effect: "Allow"
      Action:
        - kms:Decrypt
      Resource: "arn:aws:kms:#{AWS::Region}:#{AWS::AccountId}:key/${self:custom.ssmKms}"

  environment:
    SSM_FOO:
      Ref: ${self:custom.ssmFoo}

  functions:
    hello:
      handler: handler.hello
```

then the javascript would be something like this (my javascript is terrible, please don't judge 😆):

```
'use strict';

var AWS = require('aws-sdk');
var ssm = new AWS.SSM();

var ssmFoo = {
  Name: process.env.SSM_FOO,
  WithDecryption: true
}

module.exports.hello = (event, context, callback) => {
  ssm.getParameter(ssmFoo, function (err, data) {
    // bail on error
    if (err) return callback(err);
    
    // get param value from the returned object
    var foo = data.Parameter.Value;  // foo = 'bar'
    
    // do stuff with foo
  }
}
```

I don't particularly like using the callback mechanism for returning values like this...makes it very cumbersome when you need to get multiple parameters from SSM, but I haven't yet managed to figure out how to just return the value when using the `AWS.Request` class directly.

SDKs:
- [JavaScript SDK Reference](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SSM.html)
- [Python SDK Reference](https://boto3.readthedocs.io/en/latest/reference/services/ssm.html)

**Comment by candlerb** [comment]
I am in two minds now about using environment variables at all.

The good thing about using environment variables is that the settings are visible in the Lambda console.

The bad thing is that there is a temptation to change them in the Lambda console.

If you do:

1. It's not clear if they take effect immediately - but testing suggests that the lambda function is indeed restarted.

2. There is a risk that those changes are overwritten by CloudFormation the next time you do a full `sls deploy`.  Testing suggests that they aren't; but even then, your running stack is different from what CloudFormation says it should be, and that's not good.

I am therefore wondering if it's better to deploy the configuration in a file, e.g. `env_<name>.json`.  This could be deployed by serverless alongside the app, or it could be in a named S3 bucket.

The latter case means that if you update the file in the bucket, you need to restart the lambda function. Testing:

~~~
aws lambda update-function-configuration --function-name <svc>-<env>-<func>
~~~

is not by itself enough to restart the function. You need to change something, e.g.

~~~
aws lambda update-function-configuration --function-name <svc>-<env>-<func> --environment '{"Variables":{"FOO":"BAR"}}'
~~~

**Comment by tommedema** [comment]
Does serverless now create a Customer Master Key? I believe my AWS account is being charged for this but I cannot seem to delete it.

**Comment by pratri07** [comment]
Has anyone tried retrieving secrets from secrets manager using Lambda, it needs kms access. If lambda service role doesn't have kms access based on key id, it doesn't work. I mean KMS alias based permission not working to retrieve cred from secrets manager. KMS error - LATEST An error occurred (AccessDeniedException) when calling the GetSecretValue operation: Access to KMS is not allowed. If we have Lambda service role permission with kms key id, then it works fine. I don't want to put policy based on id, because rotation and per account key id will be different.

**Comment by dwyart** [comment]
Just for the record, in an IAM policy, resources are always refered by ARN, this explains what you encountered.
